autowarefoundation/autoware.universe
# feat(elevation_map_loader): add support for differential_map_loader

## Description



These PRs must be merged before these PRs.
- [PR for autoware_msg](https://github.com/autowarefoundation/autoware_msgs/pull/53)
   - modify service
- [PR for autoware_launch](https://github.com/autowarefoundation/autoware_launch/pull/268)
   - modify config for `pointcloud_map_loader`

I would like to add `differential_map_loader` support to `elevation_map_loader`.

In the default setting, `differential_map_loader` support is disabled. So if you want to use this, please change your launcher like [this](https://github.com/Shin-kyoto/autoware.universe/commit/69b1e7b9cc2139858bef8922974b43b698a0e12a).

## Related links



- [TIER IV COMPANY INTERNAL LINK](https://tier4.atlassian.net/browse/T4PB-23166)
- [PR for autoware_msg](https://github.com/autowarefoundation/autoware_msgs/pull/53)
   - modify service
- [PR for autoware_launch](https://github.com/autowarefoundation/autoware_launch/pull/268)
   - modify config for `pointcloud_map_loader`

## Tests performed



I tested with a data from Autoware tutorial. The map is divided into 20m grids ([sample-map-rosbag_split.zip](https://github.com/autowarefoundation/autoware.universe/files/10349104/sample-map-rosbag_split.zip)).

- I have confirmed that the Autoware performs the same as the current Autoware when `use_differential_load` is set false.

## Notes for reviewers



How to check

1. Use PRs below
   - [PR for autoware_msg](https://github.com/autowarefoundation/autoware_msgs/pull/53)
   - [PR for autoware_launch](https://github.com/autowarefoundation/autoware_launch/pull/268)
2. Build autoware
3. Change your launcher like [this](https://github.com/Shin-kyoto/autoware.universe/commit/69b1e7b9cc2139858bef8922974b43b698a0e12a).
4. Run command below
```bash
ros2 launch autoware_launch logging_simulator.launch.xml map_path:=$HOME/data/map/sample-map-rosbag_split pointcloud_map_file:=pointcloud_map vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit planning:=false control:=false system:=false rviz:=true sensing:=false|grep -e elevation -e map
```
5. You should get the outputs like;
```bash
❯ ros2 launch autoware_launch logging_simulator.launch.xml map_path:=$HOME/data/map/sample-map-rosbag_split pointcloud_map_file:=pointcloud_map vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit planning:=false control:=false system:=false rviz:=true sensing:=false|grep -e elevation -e map
........
[component_container-3] [INFO 1679896837.303861415] [map.pointcloud_map_loader]: Load /home/shintarotomie/data/map/sample-map-rosbag_split/pointcloud_map/89400_42300.pcd (1 out of 23)
........
[component_container_mt-1] [INFO 1679896838.782485794] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: send a request to map_loader
[component_container_mt-1] [INFO 1679896838.782555553] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: waiting response
[component_container_mt-1] [INFO 1679896838.792788277] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: concat maps
[component_container_mt-1] [INFO 1679896838.796591044] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: send a request to map_loader
[component_container_mt-1] [INFO 1679896838.796639893] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: waiting response
[component_container_mt-1] [INFO 1679896838.796889053] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: concat maps
[component_container_mt-1] [INFO 1679896838.796905702] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: finish receiving
[component_container_mt-1] [INFO 1679896838.799180362] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: receive service with pointcloud_map
[component_container_mt-1] [INFO 1679896838.799207985] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: Create elevation map from pointcloud map 
[component_container_mt-1] [INFO 1679896839.235498083] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: Finish creating elevation map. Total time: 0.433 sec
[component_container_mt-1] [INFO 1679896839.666728529] [rosbag2_storage]: Opened database '/home/shintarotomie/work/2023/0317/autoware/install/elevation_map_loader/share/elevation_map_loader/data/elevation_maps/52fe0e6043a473758c9aab905232c887a7374632dabe1099c27b22effcbd93e1/52fe0e6043a473758c9aab905232c887a7374632dabe1099c27b22effcbd93e1_0.db3' for READ_WRITE.
[component_container_mt-1] [INFO 1679896839.672484128] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: Saving elevation map successful: true
```
6. Check rviz and you should see the elevation_map like;

![image](https://user-images.githubusercontent.com/58775300/227856478-811dfc18-b9d1-4e95-b57c-d4347599b161.png)

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.


## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.


## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.


After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/
